api/capi/include/nntrainer_internal.h

```diff
@@ -77,12 +85,13 @@ typedef struct {
   std::unordered_map<std::string, ml_train_layer *> layers_map;
   ml_train_optimizer *optimizer;
   ml_train_dataset *dataset;
+  std::mutex m;
```

Comment:
Are you sure you want to add a lock to every hierarchy here?

It is usually the root of deadlocks and excessive overheads.

Reply:
In the current implementation, each object (mode, layer, dataset, and optimizer) have their own memory for the wrapper along with a shared_ptr to the actual object which is used internally. The lock in each object is to protect itself from being deleted while setting its own properties as each object has its own create(), set_property() and destroy() functions. The locks don't try to protect the shared_ptr but just the wrapper itself.

These locks are acquired together when an object is added into another (adding layer/optimizer/dataset to a model). The hierarchy of taking locks is also fixed - Model lock is always acquired first, and then layer/optimizer/dataset locks are taken.

Can you please provide some hints to reduce the usage of locks here.